### PR TITLE
rados: Make "operation" code comply to pointer passing rules

### PR DESCRIPTION
**Pull request by ansiwen**
In #581 we discovered, that we are actually violating the pointer passing rules in some places. For the write buffer of the `writeStep`s this change uses a `PtrGuard`, while the return parameters of `GetOmapStep`s are allocated in C memory.

Related: #584

### CODE REVIEW

**Comment by ansiwen** [comment]
@phlogistonjohn Ok, how about this: we switch to ptrguard by default and wait one release cycle, if someone complains. This PR I will split up so that without ptrguard tag the old ("insecure") code is used. I'm not in the mood to build another "safe" shim layer that we will remove after one release again. ;-)

File: rados/write_step.go
Comment (gman0) [reply rados/write_step.go:30]: I just noticed this comment. Does that mean PRs that would depend on this fix (e.g. https://github.com/ceph/go-ceph/pull/581) are blocked until after December's release?

Comment (phlogistonjohn) [reply rados/write_step.go:30]: No, the PRs can be merged prior to release. We simply need to discuss our options, some of the work can possibly be split up, for example.  In fact I'd prefer to merge as much as possible sooner rather than later Unfortunately, things have been slow due to illnesses and time off. Please continue to ping us on these items so that we don't forget to work on them!

**Comment by ansiwen** [comment]
Sorry, I wasn't clear enough. What I meant was: for the next release we will switch to ptrguard by default, but keep the other code. If until the following release no one complains, we can remove the non-ptrguard code.

**Comment by ansiwen** [comment]
Actually, now that I know a lot more about that whole topic, and given that the pinning API has been accepted, I would address the whole thing differently now: we can assume, that the pinning API will land before the GC becomes "moving". So before the API is there, ptrguard doesn't have to do anything. And as soon as the API is there, it can use it. The only use-case for the current implementation would be, that older versions of go-ceph would still work with some future version of go that might have a moving GC.

**Comment by phlogistonjohn** [comment]
> Actually, now that I know a lot more about that whole topic, and given that the pinning API has been accepted, I would address the whole thing differently now: we can assume, that the pinning API will land before the GC becomes "moving". So before the API is there, ptrguard doesn't have to do anything. And as soon as the API is there, it can use it. The only use-case for the current implementation would be, that older versions of go-ceph would still work with some future version of go that might have a moving GC.

Sure, I agree that it wouldn't make a lot of sense to worry about moving gc a lot given the above. To make things more "actionable" what should we do over the next couple of weeks? Break this PR up into parts, one for the ptrguard stuff and one for the other changes?  I also want to unblock PR #581.

**Comment by ansiwen** [comment]
I removed the PtrGuard related changes and just added a TODO comment for later reference.

**Review by phlogistonjohn — APPROVED** [review]
LGTM Thanks.

For a little extra context: we discussed this PR earlier today and decided that it would not be the place and time to create a hard dependency on ptrguard given that Go's GC is not a moving GC today.

**Pull request merged**
